lark/visitors.py

```diff
@@ -220,8 +221,8 @@ def transform(self, tree):
             else:
                 stack.append(x)
 
-        t ,= stack  # We should have only one tree remaining
-        return t
+        result, = stack  # We should have only one tree remaining
+        return result
```

Comment:
This is the case that is currently inconsistent. Because `self.__visit_tokens__` can cause the second if block to be skipped, it is possible to put `_Leaf_T`instances on the stack which might not be `_Return_T`. Also the specific handling of `Token` makes it unclear if this should only operate on `ParseTree` of if it can work on any arbitrary `Tree`.

Reply:
The problem here is even harder (and probably impossible) to express. `_Return_T` should in a perfect world be only the 'Top level' return type of `transform`, not of any of the intermediate steps. While those intermediate steps also often conform to that, it might not always be the case.

Reply:
That's a good point. The intermediary steps could return something that is eventually converted into a `_Return_T`. Also based on the dynamic nature of how the user functions are defined & retrieved, it likely won't be possible for `mypy` to guarantee that the final result actually is `_Result_T`.

For the first pass I wanted to see what I could do without actually changing any code. But it might be helpful have the public API methods distinct from the inner workings.

Reply:
If you can think of a way sure. But for me it seems like this is just an unfortunate truth we have to life with. (A lot of Larks internals can't be correct expressed anyway without dependent typing).